src/main/scala/com/fulcrumgenomics/umi/CollectDuplexSeqMetrics.scala

```diff
@@ -0,0 +1,540 @@
+( @arg(flag='i', doc="Input BAM file generated by GroupReadByUmi.") val input: PathToBam,
+  @arg(flag='o', doc="Prefix of output files to write.") val output: PathPrefix,
+  @arg(flag='l', doc="Optional set of intervals over which to restrict analysis.") val intervals: Option[PathToIntervals] = None,
+  @arg(flag='d', doc="Description of dataset used to label plots. Defaults to sample/library.") val description: Option[String] = None,
+  @arg(flag='a', doc="Minimum AB reads to call a tag family a 'duplex'.") val minAbReads: Int = 1,
+  @arg(flag='b', doc="Minimum BA reads to call a tag family a 'duplex'.") val minBaReads: Int = 1,
+  val generatePlots: Boolean = true // not a CLP arg - here to allow disabling of plots to speed up testing
+) extends FgBioTool with LazyLogging {
+  import CollectDuplexSeqMetrics._
+
+  // Validate inputs
+  Io.assertReadable(input)
+  Io.assertCanWriteFile(output)
+  intervals.foreach(Io.assertReadable)
+  validate(minAbReads >= 1, "min-ab-reads must be >= 1")
+  validate(minBaReads >= 0, "min-ba-reads must be >= 0")
+  validate(minBaReads <= minAbReads, "min-ab-reads must be >= min-ba-reads")
+
+  // Setup a whole bunch of counters for various things!
+  private val dsLevels               = Range.inclusive(1, 20).toArray.map(_ * 0.05)
+  private val startStopFamilyCounter = dsLevels.map(f => f -> new NumericCounter[Int]).toMap
+  private val duplexFamilyCounter    = dsLevels.map(f => f -> new SimpleCounter[Pair]).toMap
+  private val umiMetricsMap          = mutable.Map[String,UmiMetric]()
+
+  // A consensus caller used to generate consensus UMI sequences
+  private val consensusBuilder = new ConsensusCaller(errorRatePreLabeling=90.toByte, errorRatePostLabeling=90.toByte).builder()
+
+  // A Murmur3 hash used to do the downsampling of the reads by generating an int hash of the read name
+  // scaling it into the range 0-1 and then storing it a a transient attribute on the SAMRecord
+  private val hasher         = new Murmur3(42)
+  private val MaxIntAsDouble = Int.MaxValue.toDouble
+  private val HashKey        = "_P".intern()
+
+  override def execute(): Unit = {
+    // Build the iterator we'll use based on whether or not we're restricting to a set of intervals
+    val in = SamSource(input)
+    val _filteredIterator = in.iterator.filter(r => r.paired && r.mapped && r.mateMapped && r.firstOfPair && !r.secondary && !r.supplementary)
+    val iterator = intervals match {
+      case None       => _filteredIterator
+      case Some(path) =>
+        val ilist    = IntervalList.fromFile(path.toFile).uniqued(false)
+        val detector = new OverlapDetector[Interval](0,0)
+        detector.addAll(ilist.getIntervals, ilist.getIntervals)
+        _filteredIterator.filter { rec =>
+          val (start, end) = if (rec.refIndex == rec.mateRefIndex) Bams.insertCoordinates(rec) else (rec.start, rec.end)
+          detector.overlapsAny(new Interval(rec.refName, start, end))
+        }
+    }
```

Comment:
I feel like each time we do it something is slightly different - e.g. the set of filters ... using the read vs. insert cooridnates etc.